### ISSUE TITLE
malloc error when using np.divide with numpy v1.21

### ISSUE THREAD

**Issue by mcflugen** [numpy/numpy#19589]
With *numpy* 1.21 I began to see occasional errors when using *np.divide* (I also tried other *ufuncs* like *np.multiply* and saw the same errors). I believe the error only occurs when using both the *where* and *out* keywords and the *out* array has negative strides. Because the error is intermittent, I could be mistaken but that seems to be the case. 

Any help would be appreciated! Thanks!

### Reproducing code example:

This is a code sample that reproduces the problem. The key, I believe, is the negative strides in the *out* array. The error is intermittent so I may have to run the test a dozen or so times before getting an error (but usually fewer).

```python
import numpy as np

shape = (10, 6)

top = np.ones(shape, dtype=float)
bottom = np.ones(shape, dtype=float)
positive = bottom > 0.0
out = np.empty(shape, dtype=float)[::-1]

np.divide(top, bottom, where=positive, out=out)
```

### Error message:

This is the usual error message,
```
python(7282,0x108b37e00) malloc: Incorrect checksum for freed object 0x7f80e14a13d0: probably modified after being freed.
Corrupt value: 0x3ff0000000000000
python(7282,0x108b37e00) malloc: *** set a breakpoint in malloc_error_break to debug
```

but occasionally it is this,
```
python(7571,0x11282fe00) malloc: tiny_free_list_remove_ptr: Internal invariant broken (next ptr of prev): ptr=0x7ff3025189f0, prev_next=0xff00000000000000
python(7571,0x11282fe00) malloc: *** set a breakpoint in malloc_error_break to debug
```

and (very) occasionally it is just a segmentation fault,
```
[1]    7633 segmentation fault  python mytest.py
```

### NumPy/Python version information:

The error seems to only occur with *numpy* 1.21.X.

```python
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.21.1 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:36:15)
[Clang 11.1.0 ]
```

I'm on a Mac but my CI tests were also failing in a similar way on Ubuntu and Windows.

**Comment by seberg** [comment]
I can reproduce this with valgrind:
```
==61159== Invalid write of size 8
==61159==    at 0x483F803: memmove (vg_replace_strmem.c:1270)
==61159==    by 0x9D80D47: memmove (string_fortified.h:40)
==61159==    by 0x9D80D47: _contig_to_contig (lowlevel_strided_loops.c.src:375)
==61159==    by 0x9D52437: _strided_masked_wrapper_transfer_function (dtype_transfer.c:2648)
==61159==    by 0x9DCE78D: PyArray_TransferMaskedStridedToNDim (lowlevel_strided_loops.c.src:1318)
==61159==    by 0x9E02491: npyiter_copy_from_buffers (nditer_api.c:2008)
==61159==    by 0x9DFED54: npyiter_buffered_iternext (nditer_templ.c.src:312)
==61159==    by 0x9FA541A: execute_fancy_ufunc_loop (ufunc_object.c:1609)
==61159==    by 0x9FA6065: PyUFunc_GenericFunctionInternal (ufunc_object.c:2722)
==61159==    by 0x9FA79D8: ufunc_generic_fastcall (ufunc_object.c:4886)
==61159==    by 0x9FA8224: ufunc_generic_vectorcall (ufunc_object.c:5047)
==61159==    by 0x4D75BB: _PyObject_VectorcallTstate (abstract.h:118)
==61159==    by 0x4DA959: PyObject_Vectorcall (abstract.h:127)
==61159==    by 0x4DA959: call_function (ceval.c:5072)
==61159==  Address 0x52b50530 is 0 bytes after a block of size 480 alloc'd
==61159==    at 0x483877F: malloc (vg_replace_malloc.c:307)
==61159==    by 0x9CE7ACA: PyDataMem_NEW (alloc.c:246)
==61159==    by 0x9CE7C45: _npy_alloc_cache (alloc.c:92)
==61159==    by 0x9CE7C45: npy_alloc_cache (alloc.c:138)
==61159==    by 0x9D322E5: PyArray_NewFromDescr_int (ctors.c:825)
==61159==    by 0x9D32448: PyArray_NewFromDescrAndBase (ctors.c:936)
==61159==    by 0x9D32461: PyArray_NewFromDescr (ctors.c:921)
==61159==    by 0x9D33055: PyArray_Empty (ctors.c:2899)
==61159==    by 0x9DE821C: array_empty (multiarraymodule.c:1937)
==61159==    by 0x65E69A: cfunction_vectorcall_FASTCALL_KEYWORDS (methodobject.c:442)
==61159==    by 0x4D75BB: _PyObject_VectorcallTstate (abstract.h:118)
==61159==    by 0x4DA959: PyObject_Vectorcall (abstract.h:127)
==61159==    by 0x4DA959: call_function (ceval.c:5072)
==61159==    by 0x4E66C5: _PyEval_EvalFrameDefault (ceval.c:3535)
```
(continues writing 24 bytes beyond, I think. Which also for me seems to end up inside the memory Python memory arena or so.  Although, that should be random and in my case does not crash anything, but I did not try to repeat.)

Something must have gone wrong probably in the masked buffer copy :(.

**Comment by mcflugen** [comment]
@seberg It looks like you have this fixed in #19600. Thanks so much for getting to this so quickly!

Do you have an idea how long it might be for a new release with the fix? I have workarounds, so it's not super urgent for me—don't mean to be pushy, just wondering.

**Comment by seberg** [comment]
Thanks for the good report with a nice minimal reproducer. I am not sure exactly, but at the very most a few weeks, probably faster if this (or that SPARC issue that is also found just now) is painful.
